app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,197 @@
+  def handle_one_hierarchy_entry!(hierarchy)
+    entry = hierarchy.pop
+    case entry
+    when /Plain-([:\w]+)-(\d+)/
+      ar_object = $1.constantize.find($2.to_i)
+      match = hierarchy.last&.match(ASSOCIATION_RE)
+      if match
+        # callbacks logic differs between object destroyed by association, so set it here
+        # e.g. for Plans, to prevent acts_as_list to update position of each plan on deletion
+        # we want to have here a reflection with a foreign key that is part of the list scope array.
+        # see acts_as_list/active_record/acts/scope_method_definer.rb
+        # Note that if you hand-craft garbage hierarchy, it may still count as the expected foreighn key,
+        # e.g. you delete an ApplicationPlan but association is Account:account_plans or
+        # e.g. the association is for an unrelated parent object id
+        # FYI if there is no such association (e.g. a method name), nil will be returned.
+        # Another possible confusion with hand-crafter hierarchies may occur when before
+        # a Plain- entry, there is set an unrelated Hierarchy- entry, then incorrect association will be set.
+        ar_object.destroyed_by_association = match[:klass].constantize.reflect_on_association(match[:association])
+      end
+      ar_object.background_deletion_method_call
+    when ASSOCIATION_RE
+      hierarchy.concat handle_association($1.constantize.find($2.to_i), $3, entry)
+    else
+      raise ArgumentError, "Invalid entry specification: #{entry}"
     end
+    hierarchy
+  rescue ActiveRecord::RecordNotFound => exception
+    Rails.logger.warn "#{self.class} skipping object, maybe something else already deleted it: #{exception.message}"
+    []
+  rescue NameError
+    # we would be here in case of a bad crafted hierarchy entry or something else unrecoverable
+    # for example NoMethodError on nil class where retrying is pointless but logging the error is needed
+    raise DoNotRetryError, "seems like unexpectedly broken delete hierarchy entry: #{entry}"
+  end
 
-    attr_reader :main_object, :reflection, :caller_worker_hierarchy
-
-    private
-
-    def destroy_has_many_association
-      main_object.public_send("#{reflection.name.to_s.singularize}_ids").each do |associated_object_id|
-        associated_object = reflection.class_name.constantize.new
-        associated_object.id = associated_object_id
-        delete_associated_object_later(associated_object)
+  # @return a single associated object for deletion or nil if non in the association
+  def handle_association(ar_object, association, hierarchy_association_string)
+    reflection = ar_object.class.reflect_on_association(association)
+
+    case reflection.macro
+    when :has_many
+      # here we keep original hierarchy entry if we still find an associated object
+      dependent = ar_object.public_send(association).take
+      if dependent
+        dependent.destroyed_by_association = reflection
```

Comment:
Here and below in `:has_one` branch - do we need to set `destroyed_by_association` ? :thinking: 

Does it affect `hierarchy_entries_for` result in any way? (I can't see it if it does)

I understand (somewhat) why we need this information, but it seems we already fetch it in line `ar_object.destroyed_by_association = match[:klass].constantize.reflect_on_association(match[:association])` ?... :thinking: 

I guess I am missing something.

Reply:
In `hierarchy_entries_for` we fail early in case object is not destroyable instead of first deleting many or all dependents before hitting an issue with the root element. But destroyable often needs to know whether we delete by association or not.